OpenRA.Game/Primitives/Color.cs

```diff
@@ -0,0 +1,361 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Globalization;
+using OpenRA.Scripting;
+
+namespace OpenRA.Primitives
+{
+	public struct Color : IScriptBindable
+	{
+		readonly long argb;
+
+		public static Color FromArgb(int red, int green, int blue)
+		{
+			return FromArgb(byte.MaxValue, red, green, blue);
+		}
+
+		public static Color FromArgb(int alpha, int red, int green, int blue)
+		{
+			return new Color(((byte)alpha << 24) + ((byte)red << 16) + ((byte)green << 8) + (byte)blue);
+		}
+
+		public static Color FromAhsl(int alpha, float h, float s, float l)
+		{
+			// Convert from HSL to RGB
+			var q = (l < 0.5f) ? l * (1 + s) : l + s - (l * s);
+			var p = 2 * l - q;
+
+			float[] trgb = { h + 1 / 3.0f, h, h - 1 / 3.0f };
+			float[] rgb = { 0, 0, 0 };
+
+			for (var k = 0; k < 3; k++)
+			{
+				while (trgb[k] < 0) trgb[k] += 1.0f;
+				while (trgb[k] > 1) trgb[k] -= 1.0f;
+			}
+
+			for (var k = 0; k < 3; k++)
+			{
+				if (trgb[k] < 1 / 6.0f)
+					rgb[k] = p + ((q - p) * 6 * trgb[k]);
+				else if (trgb[k] >= 1 / 6.0f && trgb[k] < 0.5)
+					rgb[k] = q;
+				else if (trgb[k] >= 0.5f && trgb[k] < 2.0f / 3)
+					rgb[k] = p + ((q - p) * 6 * (2.0f / 3 - trgb[k]));
+				else
+					rgb[k] = p;
+			}
+
+			return FromArgb(alpha, (int)(rgb[0] * 255), (int)(rgb[1] * 255), (int)(rgb[2] * 255));
+		}
+
+		public static Color FromAhsl(float h, float s, float l)
+		{
+			return FromAhsl(255, h, s, l);
+		}
+
+		public static Color FromAhsv(int alpha, float h, float s, float v)
+		{
+			var ll = 0.5f * (2 - s) * v;
+			var ss = (ll >= 1 || v <= 0) ? 0 : 0.5f * s * v / (ll <= 0.5f ? ll : 1 - ll);
+			return FromAhsl(alpha, h, ss, ll);
+		}
+
+		public static Color FromAhsv(float h, float s, float v)
+		{
+			return FromAhsv(255, h, s, v);
+		}
+
+		public void ToAhsv(out int a, out float h, out float s, out float v)
```

Comment:
This was copied directly from `HSLColor.ToHSV` (plus addition of alpha channel).